### ISSUE TITLE
question about parsing performance / time complexity

### ISSUE THREAD

**Issue by sunaku** [bitwalker/combine#5]
Hello,

I would like to migrate from leex and yecc (Erlang's versions of the famous lex and yacc LALR parsing toolset) to this library but I'm concerned about the time complexity of top-down recursive descent parsing because my Elixir program needs to parse input files that can be several gigabytes in size. :scream:

I skimmed through this library's code and didn't find any use of backtracking, which is the main cause of exponential time complexity in recursive descent parsers, so it _appears_ that this library can determine whether a given input string (1) parses correctly or (2) has errors in linear time. :runner::dash:

Is this correct? How would you characterize the time and space complexity of the parsers generated by this library? Could they benefit from any further optimization, such as packrat memoization? :rat: (Edit: Oops, of course memoization wouldn't help when there's no backtracking! :disappointed_relieved:)

Thanks for your consideration.

**Comment by bitwalker** [comment]
@sunaku Hey! Thanks for the excellent questions!

Combine does have backtracking, but it's impact is limited to what parsers occur in a  - in the sense that the use of the `choice` or `either` combinators will attempt a parse from that point in the input until either a successful parse occurs, or none succeed, in which case the parser fails. What I mean by "semi-there" is that the outer combinator (let's say `choice`) holds on to the input it received, and it's children (the parsers representing the choices) will parse from there, but each parser in the choice will have a chance to start parsing from the beginning of the input held by `choice`. Once the `choice` parser succeeds, the next parser in the chain will pick up where the successful choice left off in the input, and the input consumed up to that point is discarded. So one can't backtrack to input that has already been successfully parsed, but using the combinators you can "test" multiple parsers on the current input.

TL;DR:

- Combine is only useful for parsing context-free grammars
- Combine can determine if an input string parses correctly or has errors only by parsing it, and this in general should be in linear time, but if you use branching combinators like `choice` or `either` at a high level, in the worst case each of those branches will be given a chance to parse the input, so the complexity is potentially exponential (each branching combinator multiplies the paths taken by the number of branches in the worst case, however if the branches themselves are shallow, this effect is negligible, but if the branches are deep, it obviously can have a significant performance impact).
- If you are careful about the use of branching combinators, you can have highly performant parsers which run in effectively linear time (though sadly I have not been able to write something which rivals the performance of a hand-written parser, but that's always the tradeoff with parser generators). I would characterize the time complexity of parsers generated via Combine as anywhere from linear to

**Comment by bitwalker** [comment]
Shit, everything I just wrote is wrong, and I accidentally closed it. Stupid mouse :/. Give me a few and I'll get my actual answer up.

**Comment by bitwalker** [comment]
@sunaku Hey, thanks for the excellent questions!
- Combine does have backtracking via branching combinators like `choice` or `either`.
- The effect of the backtracking is dependent on what each branch does, many of the parsers can tell on the first couple of characters of input whether they can succeed (so if you have limited choices it behaves more like a predictive parser than a recursive-descent parser with backtracking), and will fail immediately, moving to the next choice. The worst case in time complexity is exponential, due to the backtracking of course, but in practice it depends on what your parser is doing, it is absolutely possible to write linear-time parsers with Combine (for example many of the datetime parsers in Timex are linear-time, and the zoneinfo parser is as well).
- Combine discards input as it moves along, so one can't backtrack to input already successfully parsed, and it's not possible to know what was parsed already, so Combine is restricted to context-free grammars.
- Combine only knows whether input can be successfully parsed by parsing it, but it will fail as soon as it knows that it can't be successfully parsed.
- Space complexity is linear (the entire input is read into memory, and then matched on for parsing to prevent creating new binaries). So the space required is dependent on the size of the input

I haven't looked into implementing memoization, but I'm always looking for opportunities to improve the performance of Combine, so if you have suggestions on how I can do that, I'd be glad to look into it!

I hope that helps you out! I have tested Combine on ~2gb size files before (in fact it's easy to recreate this with the rfc2616 benchmark under `bench`), and it's performance was reasonably competitive in comparison to other parser combinator libraries out there. Parsing HTTP requests is naturally pretty trivial in comparison to a lot of tasks though, so it really depends on what you need to do with your parser. If your parsers are small, then the overall performance is ultimately limited by the Erlang runtime, if they are large and complex, then if performance is critical and Combine isn't satisfactory, another parser generator approach may be required (or hand rolling one, but that can be a real pain in the ass).

**Comment by sunaku** [comment]
Thanks for explaining! :+1: To minimize the impact of backtracking, I suppose one could carefully place branches (`choice` and `either`) as low :chart_with_downwards_trend: in the grammar :evergreen_tree: as possible. That way, the wasted effort of traversing a non-matching branch would be minimal because its subtree would be relatively shallow. :droplet: Conversely, the worst case would be a giant branch with deep subtrees at the root of the grammar. :umbrella:

**Comment by bitwalker** [comment]
Exactly right :), my general approach is to be as strict as possible with the grammar so as to reduce the number of branches and/or choices per branch, that combined with moving choices out towards the leaves of the tree as you suggested tends to work very well. Sometimes the choice is out of your hands though and that's when benchmarking can tell you whether you need to hand roll a parser to meet your performance requirements.

Let me know if you run into any issues or have suggestions on ways to improve Combine!
